### PR TITLE
Introduce config param for state version instead using a StateVersion::V0 for extrinsic root derivation

### PR DESCRIPTION
**Pull request by vedhavyas**
# Description

*Please include a summary of the changes and the related issue. Please also include relevant motivation and context,
including:*

- What does this PR do?
Adds Extrinsinc root state version as config parameter so that users can pick V0/V1 depending on the use case.
- Why are these changes needed?
Currently extrinsic root is derived using StateVersion::V0. In order to generate the extrinsic root, we would need to have access to full extrinsic data. We have a use case where, domains extrinsic root need to be derived on consensus runtime and passing all the extrinsic data through Fraud proof will exceed block limit. We are planning to use StateVersion::V1 so that we can just use extrinsic hashes if the extrinsic data is more than 32 bytes.
- How were these changes implemented and what do they affect?
Config parameter for ExtrinsicRootStateVersion can be passed through runtime and I have set it to StateVersion::V0 so this should not change the extrinsic root derivation for existing projects.

# Checklist

- [x] My PR includes a detailed description as outlined in the "Description" section above
- [ ] My PR follows the [labeling requirements](CONTRIBUTING.md#Process) of this project (at minimum one label for `T`
  required)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3784224

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3784597

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3791439

**Comment by vedhavyas** [comment]
Hey folks,

I could use some help in fixing the failing test and I'm not sure if this is related to my changes. Not to say the cliched thing but the following `cargo build --locked --profile testnet -p polkadot-test-malus --bin malus --bin polkadot-prepare-worker --bin polkadot-execute-worker` succefully builds on my mac and my rust version is `rustc 1.71.0-nightly` same. as the one used in the gitlab ci.

Thanks!

**Comment by cheme** [comment]
> Hey folks,
> 
> I could use some help in fixing the failing test and I'm not sure if this is related to my changes. Not to say the cliched thing but the following `cargo build --locked --profile testnet -p polkadot-test-malus --bin malus --bin polkadot-prepare-worker --bin polkadot-execute-worker` succefully builds on my mac and my rust version is `rustc 1.71.0-nightly` same. as the one used in the gitlab ci.
> 
> Thanks!

Will be reviewing the pr next week, but at first glance the CI error does not look related (also not a required one), probably no need to check before reviews (not a 'required' test).

**Review by bkchr — COMMENTED** [review]
Yeah. Generally I think this is fine. It will break everybodys runtime, but sadly we can not do better right know. However, I would re-export `StarteVersion` from `frame_system` to not require people to pull in `sp_core`.

File: polkadot/xcm/xcm-builder/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ xcm = { package = "staging-xcm", path = "..", default-features = false }
 xcm-executor = { package = "staging-xcm-executor", path = "../xcm-executor", default-features = false }
 sp-std = { path = "../../../substrate/primitives/std", default-features = false }
 sp-arithmetic = { path = "../../../substrate/primitives/arithmetic", default-features = false }
+sp-core = { path = "../../../substrate/primitives/core", default-features = false}
```
Comment (bkchr) [inline polkadot/xcm/xcm-builder/Cargo.toml:17]: Please revert the import of `sp-core` everywhere.

File: substrate/frame/system/src/lib.rs [hunk 2]
```diff
@@ -401,6 +415,9 @@ pub mod pallet {
 
 		/// The maximum number of consumers allowed on a single account.
 		type MaxConsumers: ConsumerLimits;
+
+		/// State verison used to derive extrinsics root.
+		type ExtrinsicsRootStateVersion: Get<sp_core::storage::StateVersion>;
```
Comment (cheme) [inline substrate/frame/system/src/lib.rs:420]: No idea yet if doable or fine, but would it be suitable to put this state version as part of RuntimeVersion (type Version) as a new field or a specific api version (as a specific api version, could default to 0 when missing so there would be no need to change existing pallets).

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3932572

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3932561

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3932573

**Comment by vedhavyas** [comment]
Apologies for the delay. The PR should be updated per review suggestions @bkchr

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3994935

**Comment by vedhavyas** [comment]
> No idea yet if doable or fine, but would it be suitable to put this state version as part of RuntimeVersion (type Version) as a new field or a specific api version (as a specific api version, could default to 0 when missing so there would be no need to change existing pallets).

@cheme you are correct, adding an `extrinsic_state_version` field to `RuntimeVersion` was indeed a good idea and feels more correct to me. I have opened another [PR](https://github.com/paritytech/polkadot-sdk/pull/1968) with this approach. I can close this PR if team agrees with the changing RuntimeVersion approach instead of adding a config parameter.

**Pull request closed without merging**